File: web/lib/team-activity.ts
import type { Outflow } from "./money-out";
export type TeamActivity = {
  current_outflow?: Outflow | null;
  id: string;
  team_id: string;
  business_date: string;
  actual_cny_handled: string;
  fee_rate: string;
  calculated_fee_idr: string;
  fee_status: "unpaid" | "paid";
  payment_date: string | null;
  updated_at: string;
};
export type TeamMovement = {
  id: string;
  team_id: string;
  business_date: string;
  movement_type: string;
  cny_amount: string;
  order_id: string | null;
  note: string | null;
};
export type TeamLedger = {
  balance_cny: string;
  received_cny: string;
  distributed_cny: string;
  adjustment_cny: string;
  movements: TeamMovement[];
};

File: api/app/schemas/team_activity.py
from datetime import date
from decimal import Decimal
from typing import Annotated, Literal
from uuid import UUID

from pydantic import AwareDatetime, BeforeValidator, Field, model_validator

from app.schemas.master_data import Input, Note, Rate, decimal_string
from app.schemas.orders import Amount

SignedAmount = Annotated[
    Decimal, BeforeValidator(decimal_string), Field(max_digits=18, decimal_places=2)
]


class MovementCreate(Input):
    team_id: UUID
    business_date: date
    movement_type: Literal["received", "distributed", "adjustment"]
    cny_amount: SignedAmount
    order_id: UUID | None = None
    note: Note | None = None

    @model_validator(mode="after")
    def direction(self):
        if self.movement_type != "adjustment" and self.cny_amount <= 0:
            raise ValueError("Received/distributed amounts must be positive")
        return self


class ActivityCreate(Input):
    team_id: UUID
    business_date: date
    actual_cny_handled: Amount
    fee_rate: Rate


class ActivityEdit(ActivityCreate):
    updated_at: AwareDatetime


class FeePayment(Input):
    payment_date: date
    updated_at: AwareDatetime
